### DISCUSSION TITLE
How to build OpenCascade with Emscripten

### DISCUSSION THREAD

**Discussion by Anyicheng2015** [donalffons/opencascade.js#71] [Q&A]
Hi,
I tried to build OpenCascade with Emscripten by myself, but failed.
1. Download a tagged commit from the OpenCascade git server.
2. Compile the OpenCascade library using the Emscripten compiler.

The root cause is that I don't know how to build FreeType with Emscripten.
Or is there any way to skip Freetype?

My practice is:
1. Download source code, use cmake and Emscripten tool chain to generate code.
2. Enter the result folder and tried to build, get error. "fatal error: 'ft2build.h' file not found
![occerror](https://user-images.githubusercontent.com/20434372/124702470-f52a3c00-df22-11eb-93b0-dd5f93da91f9.png)
3. I tried to take two solution, skip freetype or build freetype with Emscripten
Remove lines below, opencascade-7.5.0\CMakeLists.txt, but not work:

> if (USE_FREETYPE)
>   message (STATUS "Info: Freetype is used by OCCT")
>   OCCT_INCLUDE_CMAKE_FILE ("adm/cmake/freetype")
> else()
>   OCCT_CHECK_AND_UNSET_GROUP ("3RDPARTY_FREETYPE")
>   OCCT_CHECK_AND_UNSET ("3RDPARTY_FREETYPE_INCLUDE_DIR_freetype2")
>   OCCT_CHECK_AND_UNSET ("3RDPARTY_FREETYPE_INCLUDE_DIR_ft2build")
>   OCCT_CHECK_AND_UNSET ("INSTALL_FREETYPE")
> endif()

Tried to build freetype, but get errors:
BROTLIDEC_INCLUDE_DIRS-NOTFOUND
BROTLIDEC_LIBRARIES-NOTFOUND
BZIP2_INCLUDE_DIR-NOTFOUND
BZIP2_LIBRARY_RELEASE-NOTFOUND
Search the internet, I didn't find what is BROTLIDEC, maybe due to that I'm not familiar with FreeType.

![failed2buildft](https://user-images.githubusercontent.com/20434372/124702857-c6f92c00-df23-11eb-8844-7f5b925885ad.png)

Could you help me?
Do you have any suggestions?

Thanks,
Andy

**Comment by donalffons** [comment]
Hey there,

try to add [`-sUSE_FREETYPE=1`](https://github.com/emscripten-core/emscripten/blob/main/src/settings.js#L1410) in your Emscripten build settings. Alternatively, you could try to  exclude everything that depends on Freetype from the build (if you don't need it). I think that's mostly the [`Font`](https://github.com/Open-Cascade-SAS/OCCT/tree/master/src/Font) package.

Also, OpenCascade has a [`WebGL`](https://github.com/Open-Cascade-SAS/OCCT/tree/master/samples/webgl) example, which you can follow.

**Comment by Anyicheng2015** [comment]
Thank you! 
I tried to add  `-sUSE_FREETYPE=1` into Cmake settings. Seems not work.
Is this the right way? I'm quite new with CMake.
![image](https://user-images.githubusercontent.com/20434372/124921988-9a303c00-e02b-11eb-988c-854f1d0f5271.png)

**Reply by donalffons** [reply]
Hmm, no, that's not correct, I think.
I haven't used the CMake GUI before, but the values in that list are not passed on to Emscripten.

As a starting point, you might be able to go ahead and edit OpenCascade's `CMakeLists.txt` and manually add `set_target_properties` there. Inside that call, you can add `COMPILE_FLAGS` and `LINK_FLAGS`. `-sUSE_FREETYPE=1` is a link flag.

i.e.
```cmake
set_target_properties(sometarget PROPERTIES LINK_FLAGS "-sUSE_FREETYPE=1")
```

Have a look at [this SO issue](https://stackoverflow.com/questions/63925320/how-do-i-pass-emcc-options-through-emcmake-cmake-emmake-make).

EDIT: However, you might be able to change OCCT's CMake flag `USE_FREETYPE` (that's just named identical to Emscripten's flag) to false from that list, maybe?! In that case, OCCT should skip all packages depending on Freetype?! I'm just guessing...

**Comment by Anyicheng2015** [comment]
Hi Sebastian,
I tried to turn off the option freetype in occt-1f3c505\CMakeLists.txt:
![image](https://user-images.githubusercontent.com/20434372/125150937-03bc6180-e176-11eb-8a82-b8157d802b48.png)
But there still some error, seems freetype cannot be skipped:
![image](https://user-images.githubusercontent.com/20434372/125150963-47af6680-e176-11eb-85d8-f6798c27e3e3.png)

Could you tell me the way you build opencasecade to webassembly?
My OS is windows.

Thanks,
Andy

**Reply by donalffons** [reply]
In the current version of OpenCascade.js, I don't use the CMake based build system any more. Instead I use a custom build system...

If you want to use CMake (which is certainly recommended in most cases), you might want to have a look at [OpenCascade.js v0.1.19](https://github.com/donalffons/opencascade.js/tree/v0.1.19), which did use OpenCascade's CMake based build system.
The most relevant bits are

- [this Dockerfile](https://github.com/donalffons/opencascade.js/blob/v0.1.19/Dockerfile), which installs Emscripten and some additional tools.
- [make.py](https://github.com/donalffons/opencascade.js/blob/v0.1.19/make.py), which contains all the relevant build steps

The build happens in this order
1. [Download OCCT](https://github.com/donalffons/opencascade.js/blob/v0.1.19/make.py#L136)
2. [Extract OCCT and apply some patches](https://github.com/donalffons/opencascade.js/blob/v0.1.19/make.py#L143) (patches are probably not neccesary in your case)
3. [Download and extract regal](https://github.com/donalffons/opencascade.js/blob/v0.1.19/make.py#L159) (maybe also not necessary)
4. [Download and extract freetype](https://github.com/donalffons/opencascade.js/blob/v0.1.19/make.py#L168)
5. [Download and extract fontconfig](https://github.com/donalffons/opencascade.js/blob/v0.1.19/make.py#L177)
6. [Generate and build the JS bindings](https://github.com/donalffons/opencascade.js/blob/v0.1.19/make.py#L232) (you don't need this step or might want to use a simplified approach - depending on what you want to do)
7. [Configure OpenCascade via CMake](https://github.com/donalffons/opencascade.js/blob/v0.1.19/make.py#L262)
8. [Build OpenCascade via make](https://github.com/donalffons/opencascade.js/blob/v0.1.19/make.py#L281)

Hope that helps!

**Reply by mengxyokok** [reply]
Thank you, it's helpful for me!
